Processing data

kennankole opened the pull request:
## Data Processing
 * Created the `User`, `Comment`, `Like`, and `Post` models corresponding to all the entities of the `ER` diagram
 * Set up associations between models.
## Created custom methods 
### `User` Model
* Created `recent_posts` method that returns the 3 most recent posts for a given user.
* Created `update_post_counter` method  that updates the posts counter for a user.
* Created `recent_comments` method which returns the 5 most recent comments for a given post.
### `Post` Model
* Created `update_comment_counter` method that updates the comments counter for a post.
* Created `update_likes_counter` method that updates the likes counter for a post.

kennankole committed:
Created all model classes for all the entities in the ERD diagram
Set up all the associations between models

kennankole committed:
Added the following methods in the User and Post models.

  Users
    - created recent_posts method that returns the 3 most recent posts for a given user.
    - created update_post_counter method that updates the posts counter for a user
    - created recent_comments method that returns the 5 most recent comments for a given post.
  Posts
   - created update_comment_counter method that updates the comments counter for a post.
   - created update_likes_counter method that updates the likes counter for a post.

KDlamini commented:
## Status: Invalid 🔴 

### Hi @kennankole 🙋‍♀️,  

While you made a great effort in this project, I can't review your code.

## Invalid Code Review Request
 
You have submitted a project that has a README that is not professional. Here are key points to consider when documenting your README.

- The template logo should be removed. Optionally, the Microverse logo can be replaced with your own project logo if you have one. ✖️ 
***Please remove the Microverse heading or Rename it with the project title.***
![image](https://user-images.githubusercontent.com/38088818/220602903-73525e74-c857-4461-8a63-6f1e243c7cc5.png)

- The only optional section is the FAQ section. If there's no information in the FAQ section then it should be deleted. ✔️
- All main headers from the [README template](https://github.com/microverseinc/readme-template) should be present. Even if there's no information for a section or sub-section, rather leave texts like "NA", "None" or "coming soon" than delete it. ✔️
- Future Features section should include 1-3 potential features to be added. ✔️
-  Each section description should be removed and replaced with their custom content. The section descriptions are meant to guide you in documenting your README. They are not meant to be part of the final documentation. ✖️ 

For example:
![image](https://user-images.githubusercontent.com/38088818/216551345-4cb3cf19-9d8d-47d5-95be-ac8d47de0049.png)
![image](https://user-images.githubusercontent.com/38088818/216551482-2a274f75-4946-434a-97ea-746434b0a464.png)
![image](https://user-images.githubusercontent.com/38088818/216551578-aae6a149-69c8-4a13-bb22-4639defb421c.png)

- The License suggestion should be removed. ✖️ 
![image](https://user-images.githubusercontent.com/38088818/216764735-b1f61cef-490e-4ceb-9bc2-17b2909f1484.png)

- There should not be any blank sections that do not have content. ✔️ 

- There should not be any broken links in the README file. ✔️ 
- Delete the live demo section if the application is not deployed. If project requirements say to deploy the project, the section should be restored and the link updated. ✖️ 

![image](https://user-images.githubusercontent.com/38088818/216765466-fcad5ef6-30a6-4872-9fdc-e31a11093558.png)

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏
 
Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_Invalid Code Review Request **does not count** into the code reviews limit._

kennankole committed:
Updating the README.md file according to PR comments

codepantha commented on app/models/post.rb:
```diff
@@ -0,0 +1,22 @@
+class Post < ApplicationRecord
+  validates :title, presence: true
+  validates :text, presence: true
+  validates :comment_counter, presence: false
+  validates :likes_counter, presence: false
```
- [ ] As per requirement, please implement a method that returns the 5 most recent comments for a given post.

codepantha commented on app/models/user.rb:
```diff
@@ -0,0 +1,25 @@
+    posts.order(created_at: :desc).limit(3)
+  end
+
+  def update_post_counter
+    update(post_counter: posts.count)
+  end
```
- [ ] According to the project requirements, this method to update the posts counter for a user should be in the `Post` model instead of here. Please rectify this 👍 

**HINT:** In the `Post` model, you can take advantage of the relationship between `author` and `posts` like this 👇
```ruby
def update_post_counter
  author.increment!(:posts_counter)
end
```

kennankole replied:
Thanks alot @codepantha

codepantha commented on app/models/post.rb:
```diff
@@ -0,0 +1,22 @@
+  # retrieve all users who have commented on a post
+  has_many :commenters, through: :comments, source: :user
+
+  def update_comment_counter
+    update(comment_counter: comments.count)
+  end
+
+  def update_likes_counter
+    update(likes_counter: likes.count)
+  end
+end
```
- [ ] According to the project requirements, this method to update the comments counter should be in the `Comment` model. Please rectify this 👍 
**HINT:** In the `Comment` model, you can take advantage of the relationship between `post` and `comments` like this 👇
```ruby
def update_comments_counter
  post.increment!(:comments_counter)
end
```
- [ ] In like manner, the `#update_likes_counter` method would be better placed in the `Like` model. Kindly implement this requirement.

codepantha commented on app/models/user.rb:
```diff
@@ -0,0 +1,25 @@
+  validates :photo, presence: true
+  validates :post_counter, presence: false
+
+  has_many :comments
+  has_many :likes
+  has_many :posts
```
- [ ] The column on the comments, likes and posts tables which helps to implement the `has-many` relationship between them( the models) and `User`, is `author_id` as you can see in the attached ERD diagram below 👇 

I noticed that you made changes to your schema to be so that the foreign keys on the `comments`, `posts`, and `likes` tables are `user_id` instead of `author_id`. I highly recommend keeping your database tables and columns consistent with the ERD diagram below.

**HINT:** You can set up your model relationships this way 👇 
```ruby
`# app/models/user.rb`
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

Then in the `comment`, `like`, and `post` model, do this 👇 
```ruby
belongs_to :author, class_name: 'User'
```

![image](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png?raw=true)

codepantha reviewed (changes requested):
# STATUS: CHANGES REQUIRED ♻️ 

## Hi team 👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

<img src="https://media1.giphy.com/media/rZ10JtoPRv1Ly/giphy.gif"/>

## Highlights 🎯 
✔️ Descriptive PR
✔️ Added relationship between models 📌 
✔️ Used GitFlow
✔️ No linter errors

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

kennankole committed:
Making changes according to PR comments
https://github.com/kennankole/my-blog-app/pull/3#:~:text=No%20linter%20errors-,Required%20Changes,-Check%20the%20comments

bobb-Rob reviewed (approved):
## STATUS: APPROVED ✔️ 🍾 🍾 🍾 🥳 🥳 🥳 

Hi @kennankole, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

![](https://camo.githubusercontent.com/2c67944c14f5c74b9b607df18d73efb0c423dd148b3e10cb75b445a00220332e/68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f5061536e555a573145657970756c734f77642f67697068792e676966)

## Highlights
✅ No Linter error
✅ Correct use of Gitflow
✅ Very descriptive Pull Request(PR)
✅ Neat Commit messages

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
